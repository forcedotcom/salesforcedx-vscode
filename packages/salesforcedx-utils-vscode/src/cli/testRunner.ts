import * as path from 'path';
import * as pathExists from 'path-exists';
import { mkdir } from 'shelljs';

export class TestRunner {
  public getTempFolder(vscodePath: string, testType: string) {
    const dirPath = path.join(
      vscodePath,
      '.sfdx',
      'tools',
      'testresults',
      testType
    );

    if (!pathExists.sync(dirPath)) {
      mkdir('-p', dirPath);
    }
    return dirPath;
  }
}
