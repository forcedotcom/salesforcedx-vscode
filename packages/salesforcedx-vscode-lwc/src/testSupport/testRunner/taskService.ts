/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import * as vscode from 'vscode';

interface SfdxTaskDefinition extends vscode.TaskDefinition {
  sfdxTaskId: string;
}

/**
 * A wrapper over vscode.Task that emits events during task lifecycle
 */
export class SfdxTask {
  private task: vscode.Task;
  private taskExecution?: vscode.TaskExecution;
  public onDidStart: vscode.Event<SfdxTask>;
  public onDidEnd: vscode.Event<SfdxTask>;

  private onDidStartEventEmitter: vscode.EventEmitter<SfdxTask>;
  private onDidEndEventEmitter: vscode.EventEmitter<SfdxTask>;
  constructor(task: vscode.Task) {
    this.task = task;
    this.onDidStartEventEmitter = new vscode.EventEmitter<SfdxTask>();
    this.onDidEndEventEmitter = new vscode.EventEmitter<SfdxTask>();
    this.onDidStart = this.onDidStartEventEmitter.event;
    this.onDidEnd = this.onDidEndEventEmitter.event;
  }

  public notifyStartTask() {
    this.onDidStartEventEmitter.fire(this);
  }

  public notifyEndTask() {
    this.onDidEndEventEmitter.fire(this);
  }

  public async execute() {
    this.taskExecution = await vscode.tasks.executeTask(this.task);
    return this;
  }

  public terminate() {
    this.taskExecution!.terminate();
    this.dispose();
  }

  public dispose() {
    this.onDidStartEventEmitter.dispose();
    this.onDidEndEventEmitter.dispose();
  }
}

/**
 * Task service for creating vscode.Task
 */
class TaskService {
  private createdTasks: Map<string, SfdxTask>;

  constructor() {
    this.createdTasks = new Map();
  }

  /**
   * Register task service with extension context
   * @param context extension context
   */
  public registerTaskService(context: vscode.ExtensionContext) {
    const handleDidStartTask = vscode.tasks.onDidStartTask(
      taskStartEvent => {
        const { execution } = taskStartEvent;
        const { definition } = execution.task;
        const { sfdxTaskId } = definition;
        if (sfdxTaskId) {
          const foundTask = this.createdTasks.get(sfdxTaskId);
          if (foundTask) {
            foundTask.notifyStartTask();
          }
        }
      },
      null,
      context.subscriptions
    );

    const handleDidEndTask = vscode.tasks.onDidEndTask(
      taskEndEvent => {
        const { execution } = taskEndEvent;
        const { definition } = execution.task;
        const { sfdxTaskId } = definition;
        if (sfdxTaskId) {
          const foundTask = this.createdTasks.get(sfdxTaskId);
          if (foundTask) {
            foundTask.notifyEndTask();
            this.createdTasks.delete(sfdxTaskId);
            foundTask.dispose();
          }
        }
      },
      null,
      context.subscriptions
    );
    return vscode.Disposable.from(handleDidStartTask, handleDidEndTask);
  }

  /**
   * Create a vscode.Task instance
   * @param taskId a unique task id
   * @param taskName localized task name
   * @param taskScope task scope
   * @param cmd command line executable
   * @param args command line arguments
   */
  public createTask(
    taskId: string,
    taskName: string,
    taskScope: vscode.WorkspaceFolder | vscode.TaskScope,
    cmd: string,
    args: Array<string | vscode.ShellQuotedString>
  ): SfdxTask {
    const taskDefinition: SfdxTaskDefinition = {
      type: 'sfdxLwcTest',
      sfdxTaskId: taskId
    };
    const taskSource = 'SFDX';
    const taskShellExecution = new vscode.ShellExecution(cmd, args);
    const task = new vscode.Task(
      taskDefinition,
      taskScope,
      taskName,
      taskSource,
      taskShellExecution
    );
    task.presentationOptions.clear = true;

    const sfdxTask = new SfdxTask(task);
    this.createdTasks.set(taskId, sfdxTask);
    return sfdxTask;
  }
}
export const taskService = new TaskService();
