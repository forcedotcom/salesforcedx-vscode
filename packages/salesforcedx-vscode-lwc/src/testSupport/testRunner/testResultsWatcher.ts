/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { getTestResultsFolder } from '@salesforce/salesforcedx-utils-vscode';
import * as fs from 'fs';
import * as path from 'path';
import * as vscode from 'vscode';
import { lwcTestIndexer } from '../testIndexer';
import { TestExecutionInfo } from '../types';

/**
 * Test result watcher to watch for creating/updating test results,
 * and update test indexer.
 */
class TestResultsWatcher implements vscode.Disposable {
  private fileSystemWatchers = new Map<string, vscode.FileSystemWatcher>();
  private disposables: vscode.Disposable[] = [];

  /**
   * Register test result watcher with extension context
   * @param extensionContext extension context
   */
  public register(extensionContext: vscode.ExtensionContext) {
    extensionContext.subscriptions.push(this);
  }

  /**
   * Determine the test result output folder. It should be under
   * .sfdx/tools/testresults/lwc of the workspace folder of the test
   * @param workspaceFolder workspace folder of the test
   * @param testExecutionInfo test execution info
   */
  public getTempFolder(
    workspaceFolder: vscode.WorkspaceFolder,
    testExecutionInfo: TestExecutionInfo
  ) {
    const { testType } = testExecutionInfo;
    const workspaceFsPath = workspaceFolder.uri.fsPath;
    return getTestResultsFolder(workspaceFsPath, testType);
  }

  /**
   * Start file watchers for test results if needed.
   * The file watchers will read test result file and update test indexer.
   * @param outputFilePath Jest test results output path
   */
  public watchTestResults(outputFilePath: string) {
    const outputFileFolder = path.dirname(outputFilePath);
    let fileSystemWatcher = this.fileSystemWatchers.get(outputFileFolder);
    if (!fileSystemWatcher) {
      const outputFileExtname = path.extname(outputFilePath);
      const testResultsGlobPattern = path
        .join(outputFileFolder, `*${outputFileExtname}`)
        .replace(/\\/g, '/');
      fileSystemWatcher = vscode.workspace.createFileSystemWatcher(
        testResultsGlobPattern
      );
      fileSystemWatcher.onDidCreate(testResultsUri => {
        this.updateTestResultsFromTestResultsJson(testResultsUri);
      });

      fileSystemWatcher.onDidChange(testResultsUri => {
        this.updateTestResultsFromTestResultsJson(testResultsUri);
      });
      this.fileSystemWatchers.set(outputFileFolder, fileSystemWatcher);
      this.disposables.push(fileSystemWatcher);
    }
  }

  private updateTestResultsFromTestResultsJson(testResultsUri: vscode.Uri) {
    try {
      const { fsPath: testResultsFsPath } = testResultsUri;
      const testResultsJSON = fs.readFileSync(testResultsFsPath, {
        encoding: 'utf8'
      });
      const testResults = JSON.parse(testResultsJSON);
      lwcTestIndexer.updateTestResults(testResults);
    } catch (error) {
      console.error(error);
    }
  }

  public dispose() {
    while (this.disposables.length) {
      const disposable = this.disposables.pop();
      if (disposable) {
        disposable.dispose();
      }
    }
  }
}
export const testResultsWatcher = new TestResultsWatcher();
