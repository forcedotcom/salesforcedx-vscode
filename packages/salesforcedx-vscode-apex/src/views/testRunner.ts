/*
 * Copyright (c) 2017, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { TestRunner } from '@salesforce/salesforcedx-utils-vscode/out/src/cli/';
import * as events from 'events';
import * as vscode from 'vscode';
import {
  LanguageClientStatus,
  languageClientUtils
} from '../languageClientUtils';
import { nls } from '../messages';
import { forceApexTestRunCacheService } from '../testRunCache';
import { ReadableApexTestRunExecutor } from './readableApexTestRunExecutor';
import {
  ApexTestGroupNode,
  ApexTestNode,
  ApexTestOutlineProvider,
  TestNode
} from './testOutlineProvider';
const sfdxCoreExports = vscode.extensions.getExtension(
  'salesforce.salesforcedx-vscode-core'
)!.exports;

const EmptyParametersGatherer = sfdxCoreExports.EmptyParametersGatherer;
const SfdxCommandlet = sfdxCoreExports.SfdxCommandlet;
const SfdxWorkspaceChecker = sfdxCoreExports.SfdxWorkspaceChecker;
const channelService = sfdxCoreExports.channelService;
const sfdxCoreSettings = sfdxCoreExports.sfdxCoreSettings;

export enum TestRunType {
  All,
  Class,
  Method
}

export class ApexTestRunner {
  private testOutline: ApexTestOutlineProvider;
  private eventsEmitter: events.EventEmitter;
  constructor(
    testOutline: ApexTestOutlineProvider,
    eventsEmitter?: events.EventEmitter
  ) {
    this.testOutline = testOutline;
    this.eventsEmitter = eventsEmitter || new events.EventEmitter();
    this.eventsEmitter.on('sfdx:update_selection', this.updateSelection);
  }

  public showErrorMessage(test: TestNode) {
    let testNode = test;
    let position: vscode.Range | number = test.location!.range;
    if (testNode instanceof ApexTestGroupNode) {
      if (test.contextValue === 'apexTestGroup_Fail') {
        const failedTest = test.children.find(
          testCase => testCase.contextValue === 'apexTest_Fail'
        );
        if (failedTest) {
          testNode = failedTest;
        }
      }
    }
    if (testNode instanceof ApexTestNode) {
      const errorMessage = testNode.errorMessage;
      if (errorMessage && errorMessage !== '') {
        const stackTrace = testNode.stackTrace;
        position =
          parseInt(
            stackTrace.substring(
              stackTrace.indexOf('line') + 4,
              stackTrace.indexOf(',')
            ),
            10
          ) - 1; // Remove one because vscode location is zero based
        channelService.appendLine('-----------------------------------------');
        channelService.appendLine(stackTrace);
        channelService.appendLine(errorMessage);
        channelService.appendLine('-----------------------------------------');
        channelService.showChannelOutput();
      }
    }

    if (testNode.location) {
      vscode.window.showTextDocument(testNode.location.uri).then(() => {
        this.eventsEmitter.emit('sfdx:update_selection', position);
      });
    }
  }

  public updateSelection(index: vscode.Range | number) {
    const editor = vscode.window.activeTextEditor;
    if (editor) {
      if (index instanceof vscode.Range) {
        editor.selection = new vscode.Selection(index.start, index.end);
        editor.revealRange(index); // Show selection
      } else {
        const line = editor.document.lineAt(index);
        const startPos = new vscode.Position(
          line.lineNumber,
          line.firstNonWhitespaceCharacterIndex
        );
        editor.selection = new vscode.Selection(startPos, line.range.end);
        editor.revealRange(line.range); // Show selection
      }
    }
  }

  public getTempFolder(): string {
    if (vscode.workspace && vscode.workspace.workspaceFolders) {
      const apexDir = new TestRunner().getTempFolder(
        vscode.workspace.workspaceFolders[0].uri.fsPath,
        'apex'
      );
      return apexDir;
    } else {
      throw new Error(nls.localize('cannot_determine_workspace'));
    }
  }

  public async runAllApexTests(): Promise<void> {
    const tests = Array.from(this.testOutline.testStrings.values());
    await this.runApexTests(tests, TestRunType.All);
  }

  public async runApexTests(tests: string[], testRunType: TestRunType) {
    const languageClientStatus = languageClientUtils.getStatus() as LanguageClientStatus;
    if (!languageClientStatus.isReady()) {
      if (languageClientStatus.failedToInitialize()) {
        vscode.window.showErrorMessage(languageClientStatus.getStatusMessage());
        return Promise.resolve([]);
      }
    }

    const tmpFolder = this.getTempFolder();
    const getCodeCoverage = sfdxCoreSettings.getRetrieveTestCodeCoverage();
    if (testRunType === TestRunType.Class) {
      await forceApexTestRunCacheService.setCachedClassTestParam(tests[0]);
    } else if (testRunType === TestRunType.Method) {
      await forceApexTestRunCacheService.setCachedMethodTestParam(tests[0]);
    }
    const builder = new ReadableApexTestRunExecutor(
      tests,
      getCodeCoverage,
      tmpFolder
    );
    const commandlet = new SfdxCommandlet(
      new SfdxWorkspaceChecker(),
      new EmptyParametersGatherer(),
      builder
    );
    await commandlet.run();
  }
}
