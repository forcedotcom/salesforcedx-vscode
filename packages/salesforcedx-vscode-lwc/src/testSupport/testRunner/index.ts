/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import * as fs from 'fs';
import * as path from 'path';
import * as vscode from 'vscode';
import { nls } from '../../messages';
import { telemetryService } from '../../telemetry';
import { TestRunner, TestRunType } from './testRunner';

/**
 * Get the absolute path to LWC Test runner executable, installed in an SFDX project.
 * @param sfdxProjectPath path to the root directory of an SFDX Project
 * @returns path to lwc test runner
 */
export function getLwcTestRunnerExecutable(sfdxProjectPath: string) {
  const lwcTestRunnerExecutable = path.join(
    sfdxProjectPath,
    'node_modules',
    '.bin',
    'lwc-jest'
  );
  if (fs.existsSync(lwcTestRunnerExecutable)) {
    return lwcTestRunnerExecutable;
  } else {
    const errorMessage = nls.localize('no_lwc_jest_found_text');
    console.error(errorMessage);
    vscode.window.showErrorMessage(errorMessage);
    telemetryService
      .sendException('lwc_test_no_lwc_jest_found', errorMessage)
      .catch();
  }
}

/**
 * If testUri is specified, returns the workspace folder containing the test if it exists.
 * Otherwise, return the first workspace folder if it exists.
 * @param testUri optional testUri
 */
export function getTestWorkspaceFolder(testUri?: vscode.Uri) {
  let workspaceFolder;
  if (testUri) {
    workspaceFolder = vscode.workspace.getWorkspaceFolder(testUri);
  } else {
    workspaceFolder = vscode.workspace.workspaceFolders![0];
  }
  if (workspaceFolder) {
    return workspaceFolder;
  } else {
    const errorMessage = nls.localize(
      'no_workspace_folder_found_for_test_text'
    );
    console.error(errorMessage);
    vscode.window.showErrorMessage(errorMessage);
    telemetryService
      .sendException(
        'lwc_test_no_workspace_folder_found_for_test',
        errorMessage
      )
      .catch();
  }
}

export { TestRunner, TestRunType };
