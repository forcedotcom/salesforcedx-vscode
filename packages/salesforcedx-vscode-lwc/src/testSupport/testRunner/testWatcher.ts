/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import * as vscode from 'vscode';
import { TestExecutionInfo } from '../types';
import { FORCE_LWC_TEST_WATCH_LOG_NAME } from '../types/constants';
import { SFDX_LWC_JEST_IS_WATCHING_FOCUSED_FILE_CONTEXT } from '../types/constants';
import { SfdxTask } from './taskService';
import { TestRunner, TestRunType } from './testRunner';

/**
 * Test Watcher class for watching Jest tests
 */
class TestWatcher {
  private watchedTests: Map<string, SfdxTask> = new Map();

  /**
   * Start watching tests from provided test execution info
   * @param testExecutionInfo test execution info
   */
  public async watchTest(testExecutionInfo: TestExecutionInfo) {
    const testRunner = new TestRunner(
      testExecutionInfo,
      TestRunType.WATCH,
      FORCE_LWC_TEST_WATCH_LOG_NAME
    );
    try {
      const sfdxTask = await testRunner.executeAsSfdxTask();
      if (sfdxTask) {
        const { testUri } = testExecutionInfo;
        const { fsPath } = testUri;
        sfdxTask.onDidEnd(() => {
          this.watchedTests.delete(fsPath);
          this.setWatchingContext(testUri);
        });
        this.watchedTests.set(fsPath, sfdxTask);
        this.setWatchingContext(testUri);
        return sfdxTask;
      }
    } catch (error) {
      console.error(error);
    }
  }

  /**
   * Stop watching tests from provided test execution info
   * @param testExecutionInfo test execution info
   */
  public stopWatchingTest(testExecutionInfo: TestExecutionInfo) {
    const { testUri } = testExecutionInfo;
    const { fsPath } = testUri;
    const watchTestTask = this.watchedTests.get(fsPath);
    if (watchTestTask) {
      watchTestTask.terminate();
    }
    this.watchedTests.delete(fsPath);
    this.setWatchingContext(testUri);
  }

  /**
   * Stop watching all tests.
   */
  public stopWatchingAllTests() {
    for (const [fsPath, watchTestTask] of this.watchedTests.entries()) {
      if (watchTestTask) {
        watchTestTask.terminate();
      }
      this.watchedTests.delete(fsPath);
      this.setWatchingContext(vscode.Uri.file(fsPath));
    }
  }

  /**
   * Determine if we are watching the test uri
   * @param testUri uri of the test
   */
  public isWatchingTest(testUri: vscode.Uri) {
    const { fsPath } = testUri;
    return this.watchedTests.has(fsPath);
  }

  /**
   * Execute setContext command if applicable so that start/stop watching buttons
   * display appropriately in editor/title
   * @param testUri uri of the test
   */
  public setWatchingContext(testUri: vscode.Uri) {
    if (
      vscode.window.activeTextEditor &&
      vscode.window.activeTextEditor.document.uri.fsPath === testUri.fsPath
    ) {
      vscode.commands.executeCommand(
        'setContext',
        SFDX_LWC_JEST_IS_WATCHING_FOCUSED_FILE_CONTEXT,
        this.isWatchingTest(testUri)
      );
    }
  }
}

export const testWatcher = new TestWatcher();
