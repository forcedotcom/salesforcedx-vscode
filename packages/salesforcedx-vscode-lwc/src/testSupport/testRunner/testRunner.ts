/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { escapeStrForRegex } from 'jest-regex-util';
import * as path from 'path';
import * as uuid from 'uuid';
import * as vscode from 'vscode';
import { nls } from '../../messages';
import { telemetryService } from '../../telemetry';
import { TestExecutionInfo, TestInfoKind } from '../types';
import { getLwcTestRunnerExecutable, getTestWorkspaceFolder } from './index';
import { SfdxTask, taskService } from './taskService';
import { testResultsWatcher } from './testResultsWatcher';

export const enum TestRunType {
  RUN = 'run',
  DEBUG = 'debug',
  WATCH = 'watch'
}

/**
 * Returns relative path for Jest runTestsByPath on Windows
 * or absolute path on other systems
 * @param cwd
 * @param testFsPath
 */
export function normalizeRunTestsByPath(cwd: string, testFsPath: string) {
  if (/^win32/.test(process.platform)) {
    return path.relative(cwd, testFsPath);
  }
  return testFsPath;
}

type JestExecutionInfo = {
  jestArgs: string[];
  jestOutputFilePath: string;
};

/**
 * Test Runner class for running/debugging/watching Jest tests.
 */
export class TestRunner {
  private testExecutionInfo: TestExecutionInfo;
  private testRunType: TestRunType;
  private testRunId: string;
  private logName?: string;
  /**
   * Create a test runner from test execution info.
   * @param testExecutionInfo Test Execution information
   * @param testRunType Run, Watch or Debug
   * @param logName Telemetry log name. If specified we will send command telemetry event when task finishes
   */
  constructor(
    testExecutionInfo: TestExecutionInfo,
    testRunType: TestRunType,
    logName?: string
  ) {
    this.testRunId = uuid.v4();
    this.testExecutionInfo = testExecutionInfo;
    this.testRunType = testRunType;
    this.logName = logName;
  }

  /**
   * Deterine jest command line arguments and output file path.
   * @param workspaceFolder workspace folder of the test
   */
  public getJestExecutionInfo(
    workspaceFolder: vscode.WorkspaceFolder
  ): JestExecutionInfo | undefined {
    const { testRunId, testRunType, testExecutionInfo } = this;
    const testName =
      'testName' in testExecutionInfo ? testExecutionInfo.testName : undefined;
    const { kind, testUri } = testExecutionInfo;
    const { fsPath: testFsPath } = testUri;
    const tempFolder = testResultsWatcher.getTempFolder(
      workspaceFolder,
      testExecutionInfo
    );

    const testResultFileName = `test-result-${testRunId}.json`;
    const outputFilePath = path.join(tempFolder, testResultFileName);
    // Specify --runTestsByPath if running test on individual files
    let runTestsByPathArgs: string[];
    if (kind === TestInfoKind.TEST_FILE || kind === TestInfoKind.TEST_CASE) {
      const workspaceFolderFsPath = workspaceFolder.uri.fsPath;
      runTestsByPathArgs = [
        '--runTestsByPath',
        normalizeRunTestsByPath(workspaceFolderFsPath, testFsPath)
      ];
    } else {
      runTestsByPathArgs = [];
    }
    const testNamePatternArgs = testName
      ? ['--testNamePattern', `"${escapeStrForRegex(testName)}"`]
      : [];

    let runModeArgs: string[];
    if (testRunType === TestRunType.WATCH) {
      runModeArgs = ['--watch'];
    } else {
      runModeArgs = [];
    }
    const args = [
      ...runModeArgs,
      '--json',
      '--outputFile',
      outputFilePath,
      '--testLocationInResults',
      ...runTestsByPathArgs,
      ...testNamePatternArgs
    ];
    return {
      jestArgs: args,
      jestOutputFilePath: outputFilePath
    };
  }

  /**
   * Generate shell execution info necessary for task execution
   */
  public getShellExecutionInfo() {
    const workspaceFolder = getTestWorkspaceFolder(
      this.testExecutionInfo.testUri
    );
    if (workspaceFolder) {
      const jestExecutionInfo = this.getJestExecutionInfo(workspaceFolder);
      if (jestExecutionInfo) {
        const { jestArgs, jestOutputFilePath } = jestExecutionInfo;
        const cwd = workspaceFolder.uri.fsPath;
        const lwcTestRunnerExecutable = getLwcTestRunnerExecutable(cwd);
        let cliArgs: string[];
        if (this.testRunType === TestRunType.DEBUG) {
          cliArgs = ['--debug', '--', ...jestArgs];
        } else {
          cliArgs = ['--', ...jestArgs];
        }
        if (lwcTestRunnerExecutable) {
          return {
            workspaceFolder,
            command: lwcTestRunnerExecutable,
            args: cliArgs,
            testResultFsPath: jestOutputFilePath
          };
        }
      }
    }
  }

  /**
   * Start watching test results if needed
   * @param testResultFsPath test result file path
   */
  public startWatchingTestResults(testResultFsPath: string) {
    testResultsWatcher.watchTestResults(testResultFsPath);
  }

  private getTaskName() {
    // Only run and watch uses tasks for execution
    switch (this.testRunType) {
      case TestRunType.RUN:
        return nls.localize('run_test_task_name');
      case TestRunType.WATCH:
        return nls.localize('watch_test_task_name');
      default:
        return nls.localize('default_task_name');
    }
  }

  /**
   * Create and start a task for test execution.
   * Returns the task wrapper on task creation if successful.
   */
  public async executeAsSfdxTask(): Promise<SfdxTask | undefined> {
    const shellExecutionInfo = this.getShellExecutionInfo();
    if (shellExecutionInfo) {
      const {
        command,
        args,
        workspaceFolder,
        testResultFsPath
      } = shellExecutionInfo;
      this.startWatchingTestResults(testResultFsPath);
      const taskName = this.getTaskName();
      const sfdxTask = taskService.createTask(
        this.testRunId,
        taskName,
        workspaceFolder,
        command,
        args
      );
      if (this.logName) {
        const startTime = process.hrtime();
        sfdxTask.onDidEnd(() => {
          telemetryService.sendCommandEvent(this.logName, startTime).catch();
        });
      }
      return sfdxTask.execute();
    }
  }
}
