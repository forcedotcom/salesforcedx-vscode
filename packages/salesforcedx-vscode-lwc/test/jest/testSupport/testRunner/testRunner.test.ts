/*
 * Copyright (c) 2023, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { getTestNamePatternArgs } from '../../../../src/testSupport/testRunner/testRunner';

describe('testRunner Unit Tests.', () => {
  describe('test getTestNamePatternArgs', () => {
    it('Should return testNamePattern if flag is included', async () => {
      const testName = 'Testing is Fun!';
      const testPatternArgs = getTestNamePatternArgs(testName);
      expect(testPatternArgs.length).toEqual(2);
      expect(testPatternArgs).toMatchSnapshot();
    });
    it('Should escape certain symbols if testNamePattern is included', async () => {
      const testName = 'Test ?$^*().[]{}|+ Symbols';
      const testNameEscaped = 'Test \\?\\$\\^\\*\\(\\)\\.\\[\\]\\{\\}\\|\\+ Symbols';
      const testPatternArgs = getTestNamePatternArgs(testName);
      expect(testPatternArgs).toContain(testNameEscaped);
      expect(testPatternArgs).toMatchSnapshot();
    });
    it('Should not escape certain symbols if testNamePattern is included', async () => {
      const testName = 'Test !@#"%&;:,<>\=~` Symbols';
      const testPatternArgs = getTestNamePatternArgs(testName);
      expect(testPatternArgs).toMatchSnapshot();
    });

  });
});