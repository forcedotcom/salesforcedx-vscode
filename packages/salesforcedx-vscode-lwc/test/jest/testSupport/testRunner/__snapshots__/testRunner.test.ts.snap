// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testRunner Unit Tests. test getTestNamePatternArgs Should escape certain symbols if testNamePattern is included 1`] = `
Array [
  "--testNamePattern",
  "Test \\\\?\\\\$\\\\^\\\\*\\\\(\\\\)\\\\.\\\\[\\\\]\\\\{\\\\}\\\\|\\\\+ Symbols",
]
`;

exports[`testRunner Unit Tests. test getTestNamePatternArgs Should not escape certain symbols if testNamePattern is included 1`] = `
Array [
  "--testNamePattern",
  "Test !@#\\"%&;:,<>=~\` Symbols",
]
`;

exports[`testRunner Unit Tests. test getTestNamePatternArgs Should return testNamePattern if flag is included 1`] = `
Array [
  "--testNamePattern",
  "Testing is Fun!",
]
`;
